samples/nrf_desktop/src/hw_interface/motion_simulated.c

```diff
@@ -0,0 +1,216 @@
+/* Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+#include <atomic.h>
+
+#include "event_manager.h"
+#include "motion_event.h"
+#include "power_event.h"
+#include "hid_event.h"
+
+#include <shell/shell.h>
+#include <shell/shell_rtt.h>
+
+#define MODULE motion
+#include "module_state_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_MOTION_LOG_LEVEL);
+
+
+enum {
+	STATE_IDLE,
+	STATE_FETCHING,
+	STATE_SUSPENDED,
+};
+
+enum {
+	COORD_X,
+	COORD_Y
+};
+
+/* Coords sequence for generated trajectory */
+const static int coords[][2] = {
+				  {-300, 0}, {-210, -210}, {0, -300},
+				  {210, -210}, {300, 0}, {210, 210},
+				  {0, 300}, {-210, 210}, {-300, 0}
+			       };
+/* Time for transition between two points from trajectory */
+const static int move_period = 1000;
+
+static u32_t start_time;
+static int x_cur, y_cur;
+static atomic_t state;
+static atomic_t connected;
+
+static void motion_event_send(s8_t dx, s8_t dy)
+{
+	struct motion_event *event = new_motion_event();
+
+	event->dx = dx;
+	event->dy = dy;
+	EVENT_SUBMIT(event);
+}
+
+static void generate_motion_event(void)
+{
+	u32_t cur_time = k_uptime_get_32();
+	bool seq_rst = false;
+	int x_new, y_new;
+
+	while (cur_time - start_time
+	       >= move_period * (ARRAY_SIZE(coords) - 1)) {
+		start_time += move_period * (ARRAY_SIZE(coords) - 1);
+		seq_rst = true;
+	}
+
+	if (seq_rst) {
+		x_new = coords[0][COORD_X];
+		y_new = coords[0][COORD_Y];
+		motion_event_send(x_new - x_cur,
+				  y_new - y_cur);
+		x_cur = x_new;
+		y_cur = y_new;
+		return;
+	}
+
+	int sect = (cur_time - start_time) / move_period;
+
+	x_new = coords[sect][COORD_X]
+		+ (coords[sect + 1][COORD_X] - coords[sect][COORD_X])
+		* ((int)cur_time - (int)start_time - sect * move_period)
+		/ move_period;
+	y_new = coords[sect][COORD_Y]
+		+ ((coords[sect + 1][COORD_Y] - coords[sect][COORD_Y])
+		* ((int)cur_time - (int)start_time - sect * move_period))
+		/ move_period;
+
+	motion_event_send(x_new - x_cur, y_new - y_cur);
+
+	x_cur = x_new;
+	y_cur = y_new;
+}
+
+static bool event_handler(const struct event_header *eh)
+{
+	if (is_hid_report_subscription_event(eh)) {
+		const struct hid_report_subscription_event *event =
+			cast_hid_report_subscription_event(eh);
+
+		if (event->report_type == TARGET_REPORT_MOUSE) {
+			static u8_t peer_count;
+
+			if (event->enabled) {
+				__ASSERT_NO_MSG(peer_count < UCHAR_MAX);
+				peer_count++;
+			} else {
+				__ASSERT_NO_MSG(peer_count > 0);
+				peer_count--;
+			}
+
+			bool new_state = (peer_count != 0);
+			bool old_state = atomic_set(&connected, new_state);
+
+			if (old_state != new_state && new_state) {
+				if (atomic_get(&state) == STATE_FETCHING) {
+					start_time = k_uptime_get_32();
+					x_cur = coords[0][COORD_X];
+					y_cur = coords[0][COORD_Y];
+					generate_motion_event();
```

Comment:
Here we execute from workqueue (cooperative) context, but probably this is not a problem because the generation is very quick and does not wait for everything, and next generation will happen after `hid_report_sent_event`